### ISSUE TITLE
No support for digest

### ISSUE THREAD

**Issue by lyudmil** [opal/opal#1661]
It looks like there is no support for digest in the Opal stdlib yet. Here's the simple example I tried:

[no_digest.txt](https://github.com/opal/opal/files/981510/no_digest.txt)

It works as expected in Ruby, but when I try to compile it with Opal I get:

[error.txt](https://github.com/opal/opal/files/981519/error.txt)

Let me know if there's anything I can do to help.

**Comment by meh** [comment]
It needs to be implemented, the simplest way to do it is to find an open source JavaScript implementation for the various algorithm present in `digest` that has a compatible license, embedding it into the file and implementing the compatible API.

Look at `stdlib/base64.rb` to see an example of this.

**Comment by elia** [comment]
Hi @lyudmil! Here are the basic steps for contributing:

- fork and clone the repo (use `--recursive` to get the submodules too)
- add the relevant spec to the [`ruby_spec` list](https://github.com/opal/opal/blob/64a92b6379b6f1894606d1232b5ddf6467708300/spec/ruby_specs#L105)
- run `bundle exec rake mspec_ruby_nodejs` until all the specs are green

As @meh wrote, a good example is [`stdlib/base64`](https://github.com/opal/opal/blob/64a92b6379b6f1894606d1232b5ddf6467708300/stdlib/base64.rb)

Let me know if you need more info or guidance (also see [the Hacking guide](https://github.com/opal/opal/blob/64a92b6379b6f1894606d1232b5ddf6467708300/HACKING.md))

**Comment by lyudmil** [comment]
Hi @elia! I was very pleasantly surprised to see you among the contributors here.

Thanks for getting back to me. I'm trying to carve out some time to work on adding support for this later this week. I'll let you know if I'm successful.

**Comment by lyudmil** [comment]
I did manage to figure out how to find time to work on this, so if you're okay with it, I'll work on this this week.

As @meh mentioned, the first step is to probably choose a JS implementations of the algorithms in `Digest`. I'm leaning towards [this one](https://github.com/h2non/jshashes), but I'm terrible at dealing with licenses. It's released under a revised BSD license (read it [here](https://github.com/h2non/jshashes/blob/master/LICENSE). Would that work for us?

**Comment by meh** [comment]
@lyudmil yeah, that works, you just have to add the `LICENSE` as a comment at the top of the file.

**Comment by lyudmil** [comment]
@elia and @meh, I think I'm getting the hang of this slowly - I got the specs for SHA1 digests to pass, added some of my own, and can see a pretty clear path forward. The thing is, the JS library that backs this is almost 2K lines of code, and having a giant `%x()` at the start of the class is bugging me. What are the standard conventions you use to deal with this sort of thing?

**Comment by meh** [comment]
@lyudmil you could always add it in the `__END__` section and eval it at the beginning.

**Comment by elia** [comment]
@lyudmil or keep it in a separate file (aptly named) and require it from `stdlib/digest.rb`, e.g. it could be put in `stdlib/digest/jshashes-v1.2.3.js`

**Comment by elia** [comment]
err… forgot to mention that in that case you also should put it in an opal module, to make the require work correctly at runtime:

```js
Opal.modules['digest/jshashes-v1.2.3'] = function(Opal){
  // jshashes code here
}
```

**Comment by lyudmil** [comment]
I'm making some progress here, but I think there's a flaw in some of the specs. Let's take `spec/ruby/library/digest/md5/digest_spec.rb` for example:

```
  it "returns a digest" do
    cur_digest = Digest::MD5.new
    cur_digest.digest().should == MD5Constants::BlankDigest

    # add something to check that the state is reset later
    cur_digest << "test"

    cur_digest.digest(MD5Constants::Contents).should == MD5Constants::Digest
    # second invocation is intentional, to make sure there are no side-effects
    cur_digest.digest(MD5Constants::Contents).should == MD5Constants::Digest

    # after all is done, verify that the digest is in the original, blank state
    cur_digest.digest.should == MD5Constants::BlankDigest
  end
```

The last assertion strikes me as odd. As I understand the docs, `Digest::MD5#<<` should be an alias of `Digest::MD5#update`, which ought to have side effects. We don't do anything to reset the digest to its original blank state, so why would we expect nothing to have changed? I can confirm that `<<` is changes the content (and therefore the digest) with `irb` as well. What am I missing here?

There are a number of specs for other digest classes that follow the same pattern and currently fail for the same reason.

**Comment by meh** [comment]
@lyudmil calling `#digest` again seems to reset the internal state.

**Comment by lyudmil** [comment]
You're right, @meh! Sorry, I hadn't read through the docs closely enough. When you call `#digest` passing an argument, it's supposed to reset the digest. When I tested it out, I was calling `#digest` without any arguments, which is why I thought there was a discrepancy.

**Comment by Mogztter** [comment]
Hello @lyudmil, do you need help to complete this task ?
I need the following methods for a project and I'm willing to help :smile:

* `Digest::SHA1#hexdigest`
* `Digest::MD5#digest`

**Comment by lyudmil** [comment]
Hi @Mogztter,

All the work I have done so far is here: https://github.com/lyudmil/opal/tree/feature/support-for-digest

I've stopped actively working on this feature because I found out Opal wouldn't work for my use case for unrelated reasons. However, the branch I have up there does have support for the hash functions you need. There are others it needs to add support for, and it needs a lot of cleanup before it can be merged, but if you want to help out I'd be happy to work on this with you and get it finished.

Let me know how I can help going forward.

**Comment by Mogztter** [comment]
Thanks for sharing.
I will let you know if I need help :wink:

**Comment by hmdne** [comment]
Nowadays, JS has native functions for some hashes https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest but they are async. We could also implement those functions in webassembly and they would be sync if the resulting code could fit in 4096 bytes

**Comment by Mogztter** [comment]
Would you accept an implementation based on https://nodejs.org/api/crypto.html#crypto_crypto (only available on Node.js until we have something for browsers)?

**Comment by hmdne** [comment]
@Mogztter That's a great idea in my opinion.
